pkg/proxy/apis/config/types.go

```diff
@@ -230,10 +226,17 @@ type KubeProxyConfiguration struct {
 
 	// nodePortAddresses is a list of CIDR ranges that contain valid node IPs. If set,
 	// connections to NodePort services will only be accepted on node IPs in one of
-	// the indicated ranges. If unset, NodePort connections will be accepted on all
-	// local IPs.
+	// the indicated ranges. If this is unset and nodePortAddressesPrimary is true,
+	// then NodePort connections will be accepted only on the node's primary IP(s). If
+	// this is unset and nodePortAddressesPrimary is false, then NodePort connections
+	// will be accepted on all local IPs.
 	NodePortAddresses []string
 
+	// nodePortAddressesPrimary indicates that NodePort service connections should
+	// only be accepted on the node's primary IP(s) as indicated by the Node object.
+	// This must be false if nodePortAddresses is non-empty.
+	NodePortAddressesPrimary bool
```

Comment:
what if instead of an additional flag that we use a special value for `NodePortAddresses` to signal this behavior? this has the benefit that simplifies validation since we no longer have to cross check. I was thinking in "abuse" 0.0.0.0/8 to mean this https://datatracker.ietf.org/doc/html/rfc6890#section-2.2.2

```
         // nodePortAddresses is a list of CIDR ranges that contain valid node IPs. If set,
	// connections to NodePort services will only be accepted on node IPs in one of
	// the indicated ranges. If unset, NodePort connections will be accepted on all
	// local IPs. If set to 0.0.0.0/8, it will use the node's primary IP(s) as indicated by
        // the Node object
       // https://datatracker.ietf.org/doc/html/rfc6890#section-2.2.2
	NodePortAddresses []string
```

Reply:
We don't have to abuse an actual CIDR value, we could just allow certain non-CIDR values. That's actually [how I wrote it at first](https://github.com/danwinship/kubernetes/commit/dde6a42b63ce03d8be8de837d612c7e358cd9c03); you'd say `--nodeport-addresses primary`, and validation was changed to allow the value to be _either_ a list of CIDRs, _or_ the literal string "primary". Then after writing it I was like "no, that's weird", and rewrote it this way... But I could switch back?

Reply:
(heh, and I forgot to add the hyphen to `--nodeport-addresses-primary` in the commit message when I changed it...)

Reply:
in this case that weirdness seems to be worth it IMHO

Reply:
if we both come up to the same solution, let's if @thockin thinks the same ... is a kube-proxy only flag 🤷

Reply:
a) Using a magic v4-style CIDR seems terrible.  Let's please not. :)

b) I kind of prefer a magic string here rather than yet-another-flag with special cross-flag mutexing.

If we went back to a special value, does it need to be per-family or is it sufficient to lock them together?

If we define a special string, let's leave room for the next time we realize we've screwed up some corner case.  e.g. `primary-node-addresses` or `node:primary-addresses`

Lastly, I don't love the word "primary" - it's not defined WRT node addresses, it's just a heuristic that kube-proxy made up.

Reply:
it is really the `nodeIP`, so we can use that instead of primary

Reply:
> If we went back to a special value, does it need to be per-family or is it sufficient to lock them together?

So again with the model of "accept only on the address(es) that the cloud LB uses", and assuming that our definition of primary and secondary node IP is correct relative to that assumption, we should be able to lock them together.

Maybe I should dig into the major cloud providers and validate those assumptions...

> Lastly, I don't love the word "primary" - it's not defined WRT node addresses, it's just a heuristic that kube-proxy made up.

As above, it's not something kube-proxy just made up, though it's possibly not a widely-used name for the concept.

Reply:
is a "magic string" still an option or do we want to use the two flags?